src/jit/codegenarm64.cpp

```diff
@@ -2848,14 +2845,14 @@ void CodeGen::genCodeForCmpXchg(GenTreeCmpXchg* treeNode)
     }
 
     // The following instruction includes a release half barrier
-    // TODO-ARM64-CQ Evaluate whether this is necessary
-    // https://github.com/dotnet/coreclr/issues/14346
     getEmitter()->emitIns_R_R_R(INS_stlxr, emitTypeSize(treeNode), exResultReg, dataReg, addrReg);
 
     getEmitter()->emitIns_J_R(INS_cbnz, EA_4BYTE, labelRetry, exResultReg);
 
     genDefineTempLabel(labelCompareFail);
 
+    instGen_MemoryBarrier(INS_BARRIER_ISH);
```

Comment:
This is inserting a barrier even in the case the compare exchange fails.  This is consistent with what I remember the older gcc intrinsics did.

Reply:
I think that is the expected behavior, even on failure it indicates that the value was different and the barrier could be relied upon in that case